src/core/Database.cpp

```diff
@@ -470,30 +470,52 @@ Database* Database::unlockFromStdin(QString databaseFilename, QString keyFilenam
     return Database::openDatabaseFile(databaseFilename, compositeKey);
 }
 
-QString Database::saveToFile(QString filePath)
+/**
+ * Save the database to a file.
+ *
+ * This function uses QTemporaryFile instead of QSaveFile due to a bug
+ * in Qt (https://bugreports.qt.io/browse/QTBUG-57299) that may prevent
+ * the QSaveFile from renaming itself when using DropBox, Drive, or OneDrive.
+ *
+ * The risk in using QTemporaryFile is that the rename function is not atomic
+ * and may result in loss of data if there is a crash or power loss at the
+ * wrong moment.
+ *
+ * @param filePath Absolute path of the file to save
+ * @param keepOld Rename the original database file instead of deleting
+ * @return error string, if any
+ */
+QString Database::saveToFile(QString filePath, bool keepOld)
 {
     KeePass2Writer writer;
-    QSaveFile saveFile(filePath);
-    if (saveFile.open(QIODevice::WriteOnly)) {
-
+    QTemporaryFile saveFile;
+    if (saveFile.open()) {
         // write the database to the file
         setEmitModified(false);
         writer.writeDatabase(&saveFile, this);
         setEmitModified(true);
 
         if (writer.hasError()) {
+            // the writer failed
             return writer.errorString();
         }
 
-        if (saveFile.commit()) {
+        saveFile.close(); // flush to disk
+
+        if (keepOld) {
+            QFile::remove(filePath + ".old");
+            QFile::rename(filePath, filePath + ".old");
+        }
+
+        QFile::remove(filePath);
+        if (saveFile.rename(filePath)) {
```

Comment:
Are you sure this is going to work? If the temp folder is on a different partition, then renaming a file means copying it. Given this is rare on Windows, I'd still see it as a potential problem, because corruption could occur during copying.
Another problem I see is that on platforms other than Windows, the temp location is often /tmp, which is world-readable (the file may not be, but you can still at least list it). The database is encrypted, but we still shouldn't move it to such a location without the user's consent. Maybe make the QTemporaryFile fix Windows-only and continue using QSaveFile on other platforms. Outside of Dropbox folders it's by far the safer option and this problem is Windows-only.

Reply:
Yah I got some education from that bug thread, good to know. I did consider the option of only implementing the temp file scheme in windows.

Another issue is if a database is saved to a non-writable directory, but the database itself is writable, this implementation will actually delete the database and return.

Reply:
That's indeed a major problem.